spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala

```diff
@@ -0,0 +1,333 @@
+
+    val embedding = LookupTable[T](vocabSize, hiddenSize)
+    val embeddingDupicate = LookupTable[T](vocabSize, hiddenSize)
+    // parameter share
+    val params1 = embedding.getParameters()
+    val params2 = embeddingDupicate.getParameters()
+    params1._1.set(params2._1)
+    params1._2.set(params2._2)
+
+    val constantValue = math.sqrt(hiddenSize)
+    val embeddingInput = MulConstant(constantValue).inputs(embedding.inputs(inNode))
+    val embeddingOutput = MulConstant(constantValue).inputs(embeddingDupicate.inputs(tarNode))
+    val encoder_outputs = encode(embeddingInput, attention_bias)
+    val outNode = decode(embeddingOutput, encoder_outputs, attention_bias)
+    Graph(Array(inNode, tarNode), outNode)
+  }
+
+  private def buildLM(): Module[T] = {
+    val inNode = Input()
+    val constantValue = math.sqrt(hiddenSize)
+    val embeddingInput = MulConstant(constantValue).inputs(
+      LookupTable[T](vocabSize, hiddenSize).inputs(inNode))
+    val outNode = decode(embeddingInput)
+    Graph(inNode, outNode)
+  }
+
+  private[nn] def encode(inputs: ModuleNode[T], attention_bias: ModuleNode[T]): ModuleNode[T] = {
+    // Prepare inputs to the layer stack by adding positional encodings and
+    // applying dropout.
+    val input2 = new EncodePositionConstant().inputs(inputs)
+    val encoder_inputs = CAddTable().inputs(inputs, input2)
+    val decoder_input_drop = if (train) {
+      val postDropOut = Dropout(1- postprocessDropout)
+      postDropOut.inputs(encoder_inputs)
+    } else encoder_inputs
+
+    encodeStack(numHiddenlayers, decoder_input_drop, attention_bias)
+  }
+
+  private[nn] def decode(targets: ModuleNode[T],
+                     encoder_outputs: ModuleNode[T] = null,
+                     attention_bias: ModuleNode[T] = null): ModuleNode[T] = {
+    val decoder_input = new TransformerPrepareDecoder().inputs(targets)
+    val decoder_self_attention_bias = new SelfAttentionBiasConstant().inputs(targets)
+
+    val decoder_input_drop = if (train) {
+      val postDropOut = Dropout(1- postprocessDropout)
+      postDropOut.inputs(decoder_input)
+    } else decoder_input
+
+    decodeStack(numHiddenlayers, decoder_input_drop,
+      decoder_self_attention_bias, encoder_outputs, attention_bias)
+  }
+
+
+  private[nn] def encodeStack(num_layers: Int,
+                              encoder_input: ModuleNode[T],
+                              attention_bias: ModuleNode[T]): ModuleNode[T] = {
+    decodeStack(num_layers, encoder_input, attention_bias, preName = "encode")
+  }
+
+  private[nn] def decodeStack(num_layers: Int,
+                              decoder_input: ModuleNode[T],
+                              decoder_self_attention_bias: ModuleNode[T],
+                              encoder_outputs: ModuleNode[T] = null,
+                              attention_bias: ModuleNode[T] = null,
+                              preName: String = "decode"): ModuleNode[T] = {
```

Comment:
what does pre here mean ?

Reply:
`preName` will be used in encode and decode to distinguish layers.